flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkBuilder.java

```diff
@@ -0,0 +1,123 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.kinesis.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.connector.base.sink.AsyncSinkBaseBuilder;
+
+import software.amazon.awssdk.services.kinesis.model.PutRecordsRequestEntry;
+
+import java.util.Properties;
+
+/**
+ * Builder to construct {@link KinesisDataStreamsSink}.
+ *
+ * <p>The following example shows the minimum setup to create a {@link KinesisDataStreamsSink} that
+ * writes String values to a Kinesis Data Streams stream named your_stream_here.
+ *
+ * <pre>{@code
+ * ElementConverter<String, PutRecordsRequestEntry> elementConverter =
+ *             KinesisDataStreamsSinkElementConverter.<String>builder()
+ *                     .setSerializationSchema(new SimpleStringSchema())
+ *                     .setPartitionKeyGenerator(element -> String.valueOf(element.hashCode()))
+ *                     .build();
+ *
+ * KinesisDataStreamsSink<String> kdsSink =
+ *                 KinesisDataStreamsSink.<String>builder()
+ *                         .setElementConverter(elementConverter)
+ *                         .setStreamName("your_stream_name")
+ *                         .build();
+ * }</pre>
+ *
+ * <p>If the following parameters are not set in this builder, the following defaults will be used:
+ *
+ * <ul>
+ *   <li>{@code maxBatchSize} will be 200
+ *   <li>{@code maxInFlightRequests} will be 16
+ *   <li>{@code maxBufferedRequests} will be 10000
+ *   <li>{@code flushOnBufferSizeInBytes} will be 64MB i.e. {@code 64 * 1024 * 1024}
+ *   <li>{@code maxTimeInBufferMS} will be 5000ms
+ *   <li>{@code failOnError} will be false
+ * </ul>
+ *
+ * @param <InputT> type of elements that should be persisted in the destination
+ */
+@PublicEvolving
+public class KinesisDataStreamsSinkBuilder<InputT>
+        extends AsyncSinkBaseBuilder<
+                InputT, PutRecordsRequestEntry, KinesisDataStreamsSinkBuilder<InputT>> {
+
+    private static final int DEFAULT_MAX_BATCH_SIZE = 200;
+    private static final int DEFAULT_MAX_IN_FLIGHT_REQUESTS = 16;
+    private static final int DEFAULT_MAX_BUFFERED_REQUESTS = 10000;
+    private static final long DEFAULT_FLUSH_ON_BUFFER_SIZE_IN_B = 64 * 1024 * 1024;
+    private static final long DEFAULT_MAX_TIME_IN_BUFFER_MS = 5000;
+    private static final boolean DEFAULT_FAIL_ON_ERROR = false;
```

Comment:
The defaults were discovered during the performance test stage. The tests were run on records of size 110B and 10KB, and the defaults were tuned so the new sink was at least as good for both cases.

Reply:
The Amazon Kinesis Data Streams [PutRecrods quota](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecords.html) has a few dimensions:
> Each PutRecords request can support up to 500 records. Each record in the request can be as large as 1 MiB, up to a limit of 5 MiB for the entire request

- Batch size (record count): 500 records
- Max record size: 1MiB
- Batch size (bytes): 5MiB

How would I use these values to setup a configuration for KDS? Does `DEFAULT_FLUSH_ON_BUFFER_SIZE_IN_B` guarantee a buffer size or is it a minimum? Could we get in a situation where the payload we are trying to send exceeds the maximum size and we spin on it?

Reply:
At the moment, if I try to persist a 2MiB record, I get the following exception: `Caused by: java.util.concurrent.CompletionException: software.amazon.awssdk.services.kinesis.model.KinesisException: 1 validation error detected: Value 'java.nio.HeapByteBuffer[pos=0 lim=2500000 cap=2500000]' at 'records.1.member.data' failed to satisfy constraint: Member must have length less than or equal to 1048576`

May I recommend that a cap is introduced for `maxBatchSize`, i.e. `maxBatchSize = Math.max(maxBatchSize, 500);`.


For the max record size limitation, I believe there is nothing the sink can do if it is given a record of 1.01MiB, it just needs to error out with a clear exception.


For the batch size (bytes), it's a bit more tricky since the base sink logic deals with "flush as soon as the collection of records are greater than `flushOnBufferSizeInBytes`", which means we could get batches of larger than 5MiB. 

We could do a `flushOnBufferSizeInBytes = Math.max(flushOnBufferSizeInBytes, 4_194_304);`.

Pro: If we don't want to bog the sink down with additional logic around sizing, this would guarantee that the final record either does not exceed this limit, or is larger than 1MiB and exceeds the max record size anyway. 

Con: The downside is that if the user had set a limit of larger than 4MiB and if the sink was able to circumstantially add a few more records to make a 5MiB payload and if it were more efficient to send that payload, we'd miss out on that gain.

Reply:
Can't we be more explicit and add:
- `maxBatchSize`
- `maxBatchBytes`
- `maxRecordSize`

> For the max record size limitation, I believe there is nothing the sink can do if it is given a record of 1.01MiB, it just needs to error out with a clear exception.

Agree sink can validate record size against maximum and throw an exception if too big. We should do the validation up front though, not rely on the AWS SDK to fail.

> For the batch size (bytes), it's a bit more tricky since the base sink logic

We might need to tweak that behaviour. The next batch to be sent would need to satisfy `(batchSize < maxBatchSize) AND (batchBytes < maxBatchBytes)`

> May I recommend that a cap is introduced

We can do this, but sometimes AWS lets you raise the quota, therefore it might be best to just allow it, since it is non default. Also the quota varies per region.

Reply:
Agreed, but we have existing variables for the other two, so we just need to add `maxRecordSize`.

Sounds good, let's do the validation upfront.

Tweaking the behaviour may be controversial... it's stated in the flip that:
`As a Sink user I’d like to specify the number of bytes for the batch of requests to be flushed (e.g. ”submit the batch after the total number of bytes in it is above 1KB”) ` 
What kind of approval would you reckon I should seek if we want to modify the base sink to do `(batchSize < maxBatchSize) AND (batchBytes < maxBatchBytes)`?

...and set the default to the numbers we mentioned already...cool.

Reply:
Going to go for `(batchSize < maxBatchSize) AND (batchBytes < maxBatchBytes)` in the base sink, along with `maxBatchBytes`.

Reply:
It has been decided that we will:
1. introduce parameters `maxBatchBytes` and `maxRecordSize` in the base sink,
2. modify the base sink behaviour to `(batchSize < maxBatchSize) AND (batchBytes < maxBatchBytes)`
3. alert immediately if the sink receives a record of size greater than `maxRecordSize` in the base sink.

Reply:
Of course, after these are done, we should set these defaults

`DEFAULT_MAX_BUFFER_SIZE_IN_B = 64 * 1024 * 1024;`
`DEFAULT_MAX_RECORD_SIZE_IN_B = 5 * 1024 * 1024;`
`DEFAULT_MAX_BATCH_SIZE = 500;`